### ISSUE TITLE
Not connecting to office 365 SMTP server

### ISSUE THREAD

**Issue by knightian** [Yelp/elastalert#3021]
When trying to use email alerting over SMTP with TLS it gives the following error

`ERROR:root:Error while running alert email: Error connecting to SMTP host: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1123)`

Settings:

```
smtp_host: "smtp.office365.com"
smtp_port: 587
smtp_ssl: true
smtp_auth_file: /home/auser/elastalert/rules/smtp_auth.txt

alert:
- email

email:
- "security@anemail.com"
```

**Comment by nsano-rururu** [comment]
Since I am not using office365, I can not confirm the operation, so I will provide only the information. I'm sorry.

「smtp_ssl: true」→「smtp_ssl: false」
If smtp_ssl: false and it doesn't work, it may not work with the current construction.
https://github.com/Yelp/elastalert/issues/627#issuecomment-231824384
>You have a typo. smtp_ssl: true instead of smtp:ssl: true. Also, try port 465 is for SSL gmail smtp, 587 is for TLS (STARTTLS, which elastalert does support too, but then you don't want smtp_ssl set)

https://github.com/Yelp/elastalert/issues/2357#issuecomment-543184845
>For anyone else having this issue, I gave up with trying to get MS working and used a gmail address on port 465. Worked pretty much out the box.
>
>Appears Elastalert is using SSL and not STARTTLS which is why it doesn't work.
>
>Would be great to get this working with MS though.

I am also interested in the following blog articles. I changed the security settings

Python+Outlookでメール送信
https://masublog.net/python/pythonoutlook/
>これで送信できました。
>僕はOutlook側でセキュリティの設定を変える必要があることを知らずに動かしてエラー出しました。
>もしエラーが出たらPOPの設定をしてください。

translation
>You can now send it.
>I ran it without knowing that I needed to change the security settings on the Outlook side and got an error.
>If you get an error, please set POP.

I am also interested in the following implementation. Maybe other settings are needed?

https://github.com/Yelp/elastalert/blob/master/elastalert/alerts.py#L479
>                if self.smtp.has_extn('STARTTLS'):
>                    self.smtp.starttls(keyfile=self.smtp_key_file, certfile=self.smtp_cert_file)

**Comment by nsano-rururu** [comment]
I ran into a similar problem with gmail. Probably a configuration problem when using gmail (smtp_host:'smtp.gmail.com', smtp_port: 587), office365 (smtp_host:'smtp.office365.com', smtp_port: 587), or a program fix I think it's either necessary.

I'm avoiding it with the following settings now.

```
from_addr: "xxxx@yahoo.co.jp"
smtp_host: "smtp.mail.yahoo.co.jp"
smtp_port: 465 
smtp_ssl: true
smtp_auth_file: "/opt/elastalert/pass/smtp_auth_user.yaml"

alert:
- email

email:
- "xxx@gmail.com"
```

**Comment by knightian** [comment]
It looks like it's because MSFT is forcing TLS 1.2 for the handshane and smtplib is not using TLS 1.2

**Comment by nsano-rururu** [comment]
In the case of port 587, the encryption method is STARTTLS, so I think that it is useless if STARTTLS is not supported.

Outgoing server (SMTP) settings
Server name: smtp.office365.com
Port: 587
Encryption method: STARTTLS

**Comment by nsano-rururu** [comment]
> It looks like it's because MSFT is forcing TLS 1.2 for the handshake and smtplib is not using TLS 1.2

Do you know what kind of code to write?
If you know, please make a pull request.

**Comment by knightian** [comment]
Yah I'm looking into it

**Comment by nsano-rururu** [comment]
> It looks like it's because MSFT is forcing TLS 1.2 for the handshake and smtplib is not using TLS 1.2

What happens if I change from "smtp_ssl: true" to "smtp_ssl: false"?

**Comment by knightian** [comment]
No error but no email is sent

**Comment by nsano-rururu** [comment]
If self.smtp_key_file and self.smtp_cert_file have no values, I feel that calling "self.smtp.starttls()" with no arguments is the correct implementation.

https://github.com/Yelp/elastalert/blob/master/elastalert/alerts.py#L479
>                if self.smtp.has_extn('STARTTLS'):
>                    self.smtp.starttls(keyfile=self.smtp_key_file, certfile=self.smtp_cert_file)

**Comment by knightian** [comment]
> If self.smtp_key_file and self.smtp_cert_file have no values, I feel that calling "self.smtp.starttls()" with no arguments is the correct implementation.

Just tried but unfortunately no change

**Comment by nsano-rururu** [comment]
In the case of Office 365, it may be necessary to use the following libraries. It's just a guess
https://pypi.org/project/O365/

**Comment by nsano-rururu** [comment]
There is no more information I can provide. I can't say that I should do my best. Currently, ElastAlert seems to be dormant, so even if it is a program problem, it is unlikely that it will be fixed.

**Comment by knightian** [comment]
Ok thanks for your help

**Comment by knightian** [comment]
@nsano-rururu

Actually, the ms-teams alert was blocking the email alert from running, once I removed the teams alert and I removed the certfile etc from the starttls function it is working. I wonder why it will not send email when ms teams alert is defined above it?

**Comment by nsano-rururu** [comment]
> @nsano-rururu
> 
> Actually, the ms-teams alert was blocking the email alert from running, once I removed the teams alert and I removed the certfile etc from the starttls function it is working. I wonder why it will not send email when ms teams alert is defined above it?

Does the implementation of ms-teams influence the implementation of sending emails?

**Comment by knightian** [comment]
Well when I put this before the email stuff it only sends a teams alert and not email

```
- "ms_teams"
ms_teams_webhook_url: "https://outlook.office.com/webhook/SCRUBBED>
ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
alert_subject: "WEVTC-#4625 Admin Account Alert"
alert_text_type: alert_text_only

alert_text: |
  <br>**WEVTC:** #{0}<br><br>
  **Target Username:** {1}<br><br>
  **Target Domain:** {2}<br><br>
  **Target Endpoint:** {3}<br><br>
  **Endpoint IP:** {4}<br><br>
  **Offending IP:** {5}<br><br>
```

**Comment by knightian** [comment]
Now elastalert is telling me `did not find expected key`ugh the alerting syntax is so damn fiddly

**Comment by nsano-rururu** [comment]
> Well when I put this before the email stuff it only sends a teams alert and not email
> 
> ```
> - "ms_teams"
> ms_teams_webhook_url: "https://outlook.office.com/webhook/SCRUBBED>
> ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
> alert_subject: "WEVTC-#4625 Admin Account Alert"
> alert_text_type: alert_text_only
> 
> alert_text: |
>   <br>**WEVTC:** #{0}<br><br>
>   **Target Username:** {1}<br><br>
>   **Target Domain:** {2}<br><br>
>   **Target Endpoint:** {3}<br><br>
>   **Endpoint IP:** {4}<br><br>
>   **Offending IP:** {5}<br><br>
> 
> alert_text_args:
> - json.columns.eventid
> - osquery.EventData.TargetUserName
> - osquery.EventData.TargetDomainName
> - json.decorations.host_hostname
> - host.ip
> - osquery.EventData.IpAddress
> ```

Does it mean that if you change smtp_host to something other than office365 and confirm that it will be sent, and then add the ms_tems setting, it will be strange?
I'm not Esper, so I can't understand everything you're saying.

**Comment by nsano-rururu** [comment]
> Now elastalert is telling me `did not find expected key` ugh the alerting syntax is so damn fiddly

Isn't it just a syntax error?

**Comment by knightian** [comment]
Most likely, but with no documentation on how to actually construct the syntax it's impossible

**Comment by nsano-rururu** [comment]
> Well when I put this before the email stuff it only sends a teams alert and not email
> 
> ```
> - "ms_teams"
> ms_teams_webhook_url: "https://outlook.office.com/webhook/SCRUBBED>
> ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
> alert_subject: "WEVTC-#4625 Admin Account Alert"
> alert_text_type: alert_text_only
> 
> alert_text: |
>   <br>**WEVTC:** #{0}<br><br>
>   **Target Username:** {1}<br><br>
>   **Target Domain:** {2}<br><br>
>   **Target Endpoint:** {3}<br><br>
>   **Endpoint IP:** {4}<br><br>
>   **Offending IP:** {5}<br><br>
> 
> alert_text_args:
> - json.columns.eventid
> - osquery.EventData.TargetUserName
> - osquery.EventData.TargetDomainName
> - json.decorations.host_hostname
> - host.ip
> - osquery.EventData.IpAddress
> ```

If you haven't written the ms teams settings and you only have email alert notifications, did you receive the alert email?

**Comment by nsano-rururu** [comment]
I personally think it's better to consider another tool

**Comment by knightian** [comment]
@nsano-rururu can you recommend another?

**Comment by nsano-rururu** [comment]
Does it mean that collecting elasticsearch alert information does not change the prerequisites?

**Comment by nsano-rururu** [comment]
The following are notifications for alerts based on Elasticsearch data.

**Watcher**

https://www.elastic.co/guide/en/kibana/7.8/watcher-ui.html

**Alerting and action settings in Kibana**

[Alerts and Actions](https://www.elastic.co/guide/en/kibana/current/managing-alerts-and-actions.html)
[Managing Alerts](https://www.elastic.co/guide/en/kibana/current/alert-management.html)
[Alert details](https://www.elastic.co/guide/en/kibana/current/alert-details.html)
[Managing Connectors](https://www.elastic.co/guide/en/kibana/current/connector-management.html)

**Open Distro for Elasticsearch**

[Alerting](https://opendistro.github.io/for-elasticsearch-docs/docs/alerting/)

**Grafana**

Perform alerting on data in Elasticsearch with Grafana
https://grafana.com/docs/grafana/latest/alerting/
https://grafana.com/docs/grafana/latest/alerting/notifications/

**Comment by nsano-rururu** [comment]
Watcher and Alerting and action settings in Kibana should be paid

**Comment by nsano-rururu** [comment]
> @nsano-rururu can you recommend another?

I don't know how to use it, so check it out for yourself.

**Comment by nsano-rururu** [comment]
If you want support, we recommend other than OSS. Don't ask OSS for support. There is not always an answer, and there is no obligation to fix bugs. It's a fact

**Comment by knightian** [comment]
I understand that thanks.

This syntax sends email, but now teams doesn;'t send anything and no error is shown.

```
alert:
- email
from_addr: "security@domain.com"

email: "person@domain.com"
alert_subject: "WEVTC-#4625 Admin Account Alert"
alert_text_type: alert_text_only

alert_text: |
  <br>**WEVTC:** #{0}<br><br>
  **Target Username:** {1}<br><br>
  **Target Domain:** {2}<br><br>
  **Target Endpoint:** {3}<br><br>
  **Endpoint IP:** {4}<br><br>
  **Offending IP:** {5}<br><br>

alert_text_args:
- json.columns.eventid
- osquery.EventData.TargetUserName
- osquery.EventData.TargetDomainName
- json.decorations.host_hostname
- host.ip
- osquery.EventData.IpAddress

- ms_teams
ms_teams_webhook_url: "https://outlook.office.com/webhook/SCRUBBED
ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
alert_subject: "WEVTC-#4625 Admin Account Alert"
alert_text_type: alert_text_only

alert_text: |
  <br>**WEVTC:** #{0}<br><br>
  **Target Username:** {1}<br><br>
  **Target Domain:** {2}<br><br>
  **Target Endpoint:** {3}<br><br>
  **Endpoint IP:** {4}<br><br>
  **Offending IP:** {5}<br><br>

alert_text_args:
- json.columns.eventid
- osquery.EventData.TargetUserName
- osquery.EventData.TargetDomainName
- json.decorations.host_hostname
- host.ip
- osquery.EventData.IpAddress
```

**Comment by nsano-rururu** [comment]
alert_subject, alert_text_type, alert_text, alert_text_args should be one in the alert rule file, why are you writing the same two?

**Comment by knightian** [comment]
> alert_subject, alert_text_type, alert_text, alert_text_args should be one in the alert rule file, why are you writing the same two?

Is it not possible to have different subject per alert rule?

**Comment by nsano-rururu** [comment]
Probably not with the current specifications

**Comment by nsano-rururu** [comment]
Also, I can't answer this without looking at the source code, but some of them don't use the target settings.

**Comment by nsano-rururu** [comment]
I really don't know if it's just a document. Hard

**Comment by knightian** [comment]
I do this and it just doesn't work either

```
alert:
alert_subject: "WEVTC-#4625 Admin Account Alert"
alert_text_type: alert_text_only

alert_text: |
  <br>**WEVTC:** #{0}<br><br>
  **Target Username:** {1}<br><br>
  **Target Domain:** {2}<br><br>
  **Target Endpoint:** {3}<br><br>
  **Endpoint IP:** {4}<br><br>
  **Offending IP:** {5}<br><br>

alert_text_args:
- json.columns.eventid
- osquery.EventData.TargetUserName
- osquery.EventData.TargetDomainName
- json.decorations.host_hostname
- host.ip
- osquery.EventData.IpAddress

- ms_teams
ms_teams_webhook_url: "https://outlook.office.com/4SCRUBBED
ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
```

**Comment by nsano-rururu** [comment]
I'm not the developer of ElastAlert. What are you looking for?

**Comment by knightian** [comment]
> I'm not the developer of ElastAlert. What are you looking for?

Just to alert to teams and email but I think it's just some silly syntax things that are tripping me up

**Comment by nsano-rururu** [comment]
> Just to alert to teams and email but I think it's just some silly syntax things that are tripping me up

I understand that it is hard to create and modify everything, even if you output it with a tool and modify it by hand

**Comment by knightian** [comment]
I think what I can do is make a rule for email and a rule for teams, that should work

**Comment by knightian** [comment]
That will also give me the flexibility to format the message appearance per alert

**Comment by nsano-rururu** [comment]
The position of "-ms_teams" is different. Must be under "alert:" as shown below

```
alert:
- ms_teams

alert_subject: "WEVTC-#4625 Admin Account Alert"
alert_text_type: alert_text_only

alert_text: |
  <br>**WEVTC:** #{0}<br><br>
  **Target Username:** {1}<br><br>
  **Target Domain:** {2}<br><br>
  **Target Endpoint:** {3}<br><br>
  **Endpoint IP:** {4}<br><br>
  **Offending IP:** {5}<br><br>

alert_text_args:
- json.columns.eventid
- osquery.EventData.TargetUserName
- osquery.EventData.TargetDomainName
- json.decorations.host_hostname
- host.ip
- osquery.EventData.IpAddress

ms_teams_webhook_url: "https://outlook.office.com/4SCRUBBED
ms_teams_alert_summary: "WEVTC-#4625 Admin Account Alert"
```

**Comment by nsano-rururu** [comment]
This issue may be due to this not being merged
[Add parameter 'smtp_ca_file', fix STARTTLS problem #2681](https://github.com/Yelp/elastalert/pull/2681)

**Comment by nsano-rururu** [comment]
This issue may be due to this not being merged
[Add parameter 'smtp_ca_file', fix STARTTLS problem #2681](https://github.com/Yelp/elastalert/pull/2681)

**Comment by knightian** [comment]
Ah yes, that is working as 2 alerts 1 rule now. Thanks you for your time/help, greatly appreciated :D

**Comment by nsano-rururu** [comment]
Do you mean that you are writing as follows now?

```
alert:
- email
- ms_teams
```

**Comment by knightian** [comment]
> Do you mean that you are writing as follows now?
> 
> ```
> alert:
> - email
> - ms_teams
> ```

Yes exactly and now it works 👍🏻

**Comment by nsano-rururu** [comment]
This issue may be due to this not being merged
[Add parameter 'smtp_ca_file', fix STARTTLS problem #2681](https://github.com/Yelp/elastalert/pull/2681)
